Upgrade konnectivity-client for GRPC connection fixes

cheftako opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test
/kind flake

#### What this PR does / why we need it:

The v0.0.19 Konnectivity client includes several
significant fixes to prevent the GRPC tunnel between
the KAS and the APIServer Network Proxy from becoming
blocked/wedged. 
Importantly it picks up the fix for https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/167.
We believe this will also fix many of the failures currently seen on
https://testgrid.k8s.io/sig-api-machinery-network-proxy#ci-kubernetes-e2e-gci-gce-network-proxy-grpc&width=5.

#### Which issue(s) this PR fixes:

Fixes #101879

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
None
```

cheftako commented:
/sig cloud-provider

cheftako commented:
/priority important-soon

cheftako commented:
/cc @liggitt

cheftako commented:
/cc @caesarxuchao

liggitt commented:
sorry, several deps got updated recently and openapi got dropped, you might need to replay the konnectivity pin/update-vendor

liggitt commented on go.mod:
```diff
@@ -42,6 +42,7 @@ require (
 	github.com/evanphx/json-patch v4.9.0+incompatible
 	github.com/fsnotify/fsnotify v1.4.9
 	github.com/go-bindata/go-bindata v3.1.1+incompatible
+	github.com/go-openapi/spec v0.20.3
```
this addition looks like a bad rebase

cheftako replied:
Showed up when I ran hack/update-vendor.sh

liggitt replied:
make sure you're rebased on latest master and `make clean` first?

cheftako replied:
```
get fetch upstream
git rebase upstream/master
make clean
./hack/pin-dependency.sh sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.15
./hack/update-vendor.sh
```
Still getting it. (go 1.16.4)

liggitt commented on vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go:
```diff
@@ -129,7 +139,12 @@ func (t *grpcTunnel) serve(c clientConn) {
-				conn.readCh <- resp.Data
+				select {
+				case conn.readCh <- resp.Data:
+				case <-time.After((time.Duration)(t.readTimeoutSeconds) * time.Second):
```
it looks like this is hit once per data packet, which seems high-frequency. `time.After` does not release resources until the time fires (https://golang.org/pkg/time/#After)

> The underlying Timer is not recovered by the garbage collector until the timer fires. If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed.

cheftako replied:
/cc @caesarxuchao

cheftako replied:
Resolved with https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/220 (Post v0.0.18)

liggitt commented on vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go:
```diff
@@ -114,6 +120,10 @@ func (t *grpcTunnel) serve(c clientConn) {
 					err:    resp.Error,
 					connid: resp.ConnectID,
 				}
+				// We should only get DialResp once. If we get the DialResp
+				// twice, we will get an exception trying to write to a closed
+				// channel.
+				close(ch)
```
what's the failure mode if this happens?

cheftako replied:
When we get the second dial response, before we've had a chance to remove the channel from the pendingDial map (talking milliseconds here) then line 119 will get an error writing to a closed channel. That should leave us with a stacktrace and kill this goroutine. That should stop also stop us from hanging the GRPC tunnel.

liggitt replied:
That will panic and crash the server, right?

cheftako replied:
Last time I checked the answer would have been no but apparently we've changed that default since last time I checked.

liggitt replied:
spoke with walter offline, I think the plan is to change from a `close(ch)` (which can panic the server) to something like this:

```go
select {
case ch <- dialResult{err: resp.Error, connid: resp.ConnectID}:
  // write to channel succeeded
default:
  // write to channel blocked, which is unexpected
  ... log nasty message ...
  return
}
```

cheftako replied:
Resolved with https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/221 (Post v0.0.18)

cheftako commented:
/hold Need to switch to the v0.0.19 version of the client which resolve @liggitt concerns.

ydp commented:
Hello, is it possible that the fix goes to 1.21 patch release as well?

aojea commented:
/test pull-kubernetes-e2e-gce-network-proxy-grpc

aojea commented:
> /test pull-kubernetes-e2e-gce-network-proxy-grpc

:clap: only one failure on the grpc job

```
Kubernetes e2e suite: [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] should be able to deny custom resource creation, update and deletion [Conformance]
```

cheftako commented:
> Hello, is it possible that the fix goes to 1.21 patch release as well?

Yes, I will certainly push to have this cherry-picked. We'll be switching to the v0.0.19 version of the client. I still need to resolve the go.mod oddity.

cheftako commented:
/test pull-kubernetes-e2e-gce-network-proxy-grpc

Jefftree commented:
@cheftako Even though it's probably a no-op, let's also bump the images for the [server](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/konnectivity-server.yaml) and [agent](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/addons/konnectivity-agent/konnectivity-agent-ds.yaml) to 0.0.19.

cheftako commented:
> @cheftako Even though it's probably a no-op, let's also bump the images for the [server](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/konnectivity-server.yaml) and [agent](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/addons/konnectivity-agent/konnectivity-agent-ds.yaml) to 0.0.19.

:100:

cheftako committed:
Upgrade konnectivity-client for GRPC connection fixes

The v0.0.19 Konnectivity client includes several
significant fixes to prevent the GRPC tunnel between
the KAS and the APIServer Network Proxy from becoming
blocked/wedged.
Importantly it picks up the fix for kubernetes-sigs/apiserver-network-proxy#167.
We believe this will also fix many of the failures currently seen on
https://testgrid.k8s.io/sig-api-machinery-network-proxy#ci-kubernetes-e2e-gci-gce-network-proxy-grpc&width=5.

cheftako commented:
/test pull-kubernetes-e2e-gce-network-proxy-grpc

cheftako commented:
/test pull-kubernetes-node-e2e-containerd

cheftako commented:
/remove-hold

liggitt commented:
/approve

liggitt commented:
/lgtm

caesarxuchao commented:
/lgtm
/approve

fedebongio commented:
/triage accepted
